### PR TITLE
Parameters, varianter and runner

### PR DESCRIPTION
**Pull request by clebergnu**
This is a collection of refactors and documentation improvements in preparation for bigger changes regarding test parameters.

The highlight is probably the separation of the base parameter classes from `varianter`, into its own module, `avocado.core.parameters` and the removal of a test "default_parameters".

Then next batch of work should include:
  * the introduction of a generic parameter mechanism
  * removal of the "avocado_default_parameters"

### CODE REVIEW

**Comment by clebergnu** [comment]
[Get rid of default_params](https://trello.com/c/9scZuNOm/585-get-rid-of-defaultparams)

File: optional_plugins/varianter_yaml_to_mux/avocado_varianter_yaml_to_mux/__init__.py [hunk 1]
```diff
@@ -22,7 +22,7 @@
 
 from six import iteritems
 
-from avocado.core import tree, exit_codes, varianter
+from avocado.core import tree, exit_codes
```
Comment (apahim) [inline optional_plugins/varianter_yaml_to_mux/avocado_varianter_yaml_to_mux/__init__.py]: ok

File: selftests/unit/test_mux.py [hunk 2]
```diff
@@ -312,7 +312,7 @@ def test_basic(self):
         self.assertNotEqual(self.params1, self.params2)
         repr(self.params1)
         str(self.params1)
-        str(varianter.AvocadoParams([], 'Unittest', [], {}))
+        str(parameters.AvocadoParams([], 'Unittest', [], {}))
```
Comment (apahim) [inline selftests/unit/test_mux.py]: ok, no actual change of behavior. The module split makes sense even if we hadn't further goals.

File: avocado/core/parameters.py [hunk 3]
```diff
@@ -149,22 +149,6 @@ def _is_abspath(path):
         else:
             return False
 
-    def __getattr__(self, attr):
-        """
-        Compatibility to old Params
-        :warning: This will be removed soon. Use params.get() instead
```
Comment (apahim) [inline avocado/core/parameters.py]: Yep, this warning was created in Apr 9, 2015 :)

File: avocado/core/runner.py [hunk 4]
```diff
@@ -506,8 +506,12 @@ def _template_to_factory(template, variant):
...
         Applies test params from variant to the test template
 
-        :param template: a test template
-        :param variant: variant to be applied
+        :param template: a test template, containing the class name,
+                         followed by parameters to the class
+        :type template: tuple
+        :param variant: variant to be applied, usually containing
+                        the keys: mux_path, variant and variant_id
+        :type variant: dict
```
Comment (apahim) [inline avocado/core/runner.py]: ok

File: avocado/core/runner.py [hunk 5]
```diff
@@ -515,25 +515,22 @@ def _template_to_factory(template, variant):
...
+                       "test %s to not to fill variants."
+                       % (variant, template))
+                raise NotImplementedError(msg)
+            params = template[1]["params"]
+            variant_id = varianter.generate_variant_id(params[0])
+            return template, {"variant": params[0],
+                              "variant_id": variant_id,
+                              "mux_path": params[1]}
+        factory = [template[0], template[1].copy()]
+        factory[1]["params"] = params
```
Comment (apahim) [inline avocado/core/runner.py]: ok

File: avocado/core/runner.py [hunk 6]
```diff
@@ -514,24 +514,27 @@ def _template_to_factory(template, variant):
...
                               "variant_id": variant_id,
-                              "mux_path": params[1]}
-        factory = [template[0], template[1].copy()]
-        factory[1]["params"] = params
-        return factory, variant
+                              "mux_path": mux_path}
+        else:
+            factory = [klass, klass_parameters.copy()]
+            factory[1]["params"] = (var, mux_path)
+            return factory, variant
```
Comment (apahim) [inline avocado/core/runner.py]: ok, nice refactor

File: avocado/core/runner.py [hunk 7]
```diff
@@ -595,6 +600,8 @@ def run_suite(self, test_suite, variants, timeout=0, replay_map=None,
             for test_template in test_suite:
                 test_template[1]["base_logdir"] = self.job.logdir
                 test_template[1]["job"] = self.job
+            if execution_order is None:
+                execution_order = self.DEFAULT_EXECUTION_ORDER
```
Comment (apahim) [inline avocado/core/runner.py]: ok

File: selftests/unit/test_mux.py [hunk 8]
```diff
@@ -312,7 +312,7 @@ def test_basic(self):
         self.assertNotEqual(self.params1, self.params2)
         repr(self.params1)
         str(self.params1)
-        str(parameters.AvocadoParams([], 'Unittest', [], {}))
+        str(parameters.AvocadoParams([], 'Unittest', []))
```
Comment (apahim) [inline selftests/unit/test_mux.py]: ok

**Comment by apahim** [comment]
`selftests/doc/test_doc_build.py` is failing on `Parameters: remove the concept of a test default parameters`:
```
docs/source/TestParameters.rst:90: ERROR: Unknown target name: "test's default params".
docs/source/TestParameters.rst:134: ERROR: Unknown target name: "test's default params".
```

**Comment by clebergnu** [comment]
@apahim nice catch wrt to the needed doc changes.  I'm wondering why I didn't catch this in the selftests...

**Comment by clebergnu** [comment]
Closing in favor of #2322

**Pull request closed without merging**